tests/system/src/test/java/test/robot/javafx/scene/TextFlowSurrogatePairInsertionIndexTest.java

```diff
@@ -0,0 +1,355 @@
+ * 3. Insertion index should be same as character index.
+ * 4. Move the cursor continously till last character and check that
+ *    character index and insertion index increase monitonically as expected.
+ *
+ * Steps for testTextFlowInsertionIndexUsingEmbeddedTextNodes()
+ * 1. Create a TextFlow. Add a Text node with text and another with emojis.
+ * 2. Move the cursor to the first character and click.
+ * 3. Insertion index should be same as character index.
+ * 4. Move the cursor continously till last character and check that
+ *    character index and insertion index increase monitonically as expected.
+ *
+ * Steps for testTextFlowInsertionIndexWhenMouseMovedOutsideText()
+ * 1. Create a TextFlow. Add a Text node with text and emojis.
+ * 2. Move the cursor to the first character and click.
+ * 3. Insertion index should be same as character index.
+ * 4. Move the cursor towards bottom of the application window check that
+ *    chracter index and insertion index are as expected.
+ * This test is implemented to test insertion index initialization
+ * when text run is not used to calculate character index.
+ *
+ * Steps for testTextFlowInsertionIndexUsingWrappedText()
+ * 1. Create a TextFlow. Add a Text node with text and emojis whose length is
+ *    more than the size of application window.
+ * 2. Move the cursor from first character to last character.
+ * 3. Character index should increase monotonically as expected.
+ * 4. Insertion index should also increase as expected along with character index.
+ */
+
+public class TextFlowSurrogatePairInsertionIndexTest {
+    static CountDownLatch startupLatch = new CountDownLatch(1);
+    static Robot robot;
+    static TextFlow textFlow;
+    static Text text;
+    static Text emoji;
+
+    static volatile Stage stage;
+    static volatile Scene scene;
+
+    static final int WIDTH = 500;
+    static final int HEIGHT = 150;
+
+    final int Y_OFFSET = 25;
+    final int X_LEADING_OFFSET = 10;
+    final int X_TRAILING_OFFSET = 40;
+
+    boolean isLeading;
+    boolean isSurrogatePair;
+    int charIndex;
+    int insertionIndex;
+
+    private void mouseClick(double x, double y) {
+        Util.runAndWait(() -> {
+            robot.mouseMove((int) (scene.getWindow().getX() + scene.getX() + x),
+                    (int) (scene.getWindow().getY() + scene.getY() + y));
+            robot.mouseClick(MouseButton.PRIMARY);
+        });
+    }
+
+    private void moveMouseOverTextFlow(int x, int y) throws Exception {
+        mouseClick(textFlow.getLayoutX() + x,
+                    textFlow.getLayoutY() + y);
+    }
+
+    private void addTwoEmojis() {
+        Util.runAndWait(() -> {
+            text = new Text("😊😇");
+            text.setFont(new Font(48));
+            textFlow.getChildren().clear();
+            textFlow.getChildren().setAll(text);
+        });
+    }
+
+    private void addMultipleEmojis() {
+        Util.runAndWait(() -> {
+            text = new Text("😊😇💙🦋🏁🔥");
```

Comment:
This case is same as the one which we saw with red heart as the emoji sequence mentioned here is also a U+FE0F Variation Selector-16 character. Since it exhibits different behavior in different platforms we decided to remove it.

I compared the insertion index calculated for these emojis with the insertion index calculated for red heart emoji. The insertion index calculated looks to be correct for the above given emojis. It is considered as 2 characters if color is also present else it is considered as single character.
For me it is giving correct caret position as well in the Monkey tester.

I found strange behavior when I tried to add heart emojis (with and without color) along with the `CLUSTERS` variable present in Monkey Tester so that it gets displayed when Rich Text option is selected.
When `CLUSTERS = "❤️❤❤️"`, emojis are displayed properly.
<img width="658" alt="image" src="https://github.com/openjdk/jfx/assets/26969459/f3de5d1d-ad8c-42a6-8284-01916dfd37a8">

When `CLUSTERS = "☝🏿☝🏿☝🏿🤦🏼‍♂️❤️❤❤️";`, following sequence is displayed.
<img width="532" alt="image" src="https://github.com/openjdk/jfx/assets/26969459/6eaa67b8-6fa1-475f-82bd-e9c48240b1c1">

This looks like a separate issue. Please let me know your thoughts on this.

Reply:
@karthikpandelu : 
1. you are right, multiple symbols and a failed test are due to lack of proper handling of grapheme clusters in fx.  I don't think we have a ticket, so created JDK-8309565 for that
2. your latest fix uncovered a problem in my RichTextArea code, so thank you!

The new code passes all my tests in the MonkeyTester and RichTextArea.

Reply:
Than you @andy-goryachev-oracle for the detailed review and feedback.